src/Rules/DeadCode/PossiblyPureFuncCallCollector.php

```diff
@@ -0,0 +1,44 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\DeadCode;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Collectors\Collector;
+use PHPStan\Reflection\ReflectionProvider;
+
+/**
+ * @implements Collector<FuncCall, array{string, int}>
+ */
+class PossiblyPureFuncCallCollector implements Collector
+{
+
+	public function __construct(private ReflectionProvider $reflectionProvider)
+	{
+	}
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	public function processNode(Node $node, Scope $scope)
+	{
+		if (!$node->name instanceof Node\Name) {
+			return null;
+		}
+
+		if (!$this->reflectionProvider->hasFunction($node->name, $scope)) {
+			return null;
+		}
+
+		$functionReflection = $this->reflectionProvider->getFunction($node->name, $scope);
```

Comment:
We should also have this condition:

```php
if (!$function->hasSideEffects()->maybe() {
    return null;
}
```

Because:

1) If it's `no()` it's already covered by CallToFunctionStatementWithoutSideEffectsRule
2) if it's `yes()` then it's a valid call and should not be reported in this rule.

The distinction between `hasSideEffects()` and `isPure` essentially boils down to function being `void` and function being marked as `@phpstan-pure` and `@phpstan-impure`.

Here we're dealing with the uncertainty so we need both of these to be `maybe()`.